galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,320 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceentry
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/galley/pkg/metadata"
+	"istio.io/istio/galley/pkg/runtime/log"
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+	"istio.io/istio/galley/pkg/runtime/processing"
+	"istio.io/istio/galley/pkg/runtime/projections/serviceentry/convert"
+	"istio.io/istio/galley/pkg/runtime/projections/serviceentry/locality"
+	"istio.io/istio/galley/pkg/runtime/projections/serviceentry/node"
+	"istio.io/istio/galley/pkg/runtime/projections/serviceentry/pod"
+	"istio.io/istio/galley/pkg/runtime/resource"
+	"istio.io/istio/pkg/mcp/snapshot"
+
+	coreV1 "k8s.io/api/core/v1"
+)
+
+var (
+	scope      = log.Scope
+	collection = metadata.IstioNetworkingV1alpha3SyntheticServiceentries.Collection
+
+	// Schema for types required to generate synthetic ServiceEntry projections.
+	Schema *resource.Schema
+)
+
+func init() {
+	b := resource.NewSchemaBuilder()
```

Comment:
Schema is not meant to be used this way (i.e. schema of sub-components). This should be a list of resource.Collections, on which you can register the handler.

Reply:
You can let the dispatcher builder passed in as part of init function to do this.

Reply:
The code to register ends up being uglier...

```golang
b.Register(metadata.K8sCoreV1Pods.Collection.String(), metadata.K8sCoreV1Pods.TypeURL.String())
```

The `Info` object isn't really all that special .... why not just build a schema from them directly?